### DISCUSSION TITLE
`visit_locations` table contains location data but frontend says "unknown"

### DISCUSSION THREAD

**Discussion by GammelSami** [shlinkio/discussions#95] [Help wanted] [answered]
#### How Shlink is set-up

* Shlink Version: 3.1.1
* PHP Version: 8.1.5
* How do you serve Shlink: Docker image
* Database engine used: MySQL
* Frontend Version: 3.7.1 (using app.shlink.io)

#### Summary

After generating location data all visits in the frontend still have "unknown" country and city. But when I look it up in the mysql database the location data is there.
(using app.shlink.io)

#### How to reproduce

1. View a short link
2. Generate location data using `shlink visit:locate`
3. View location data in frontend

**Comment by acelaya** [comment]
The rows in the visit_locations table do not mean geolocation was possible, just that it was attempted (to avoid trying to geolocate the same visit over and over, unles `--force` is provided).

If the front-end still shows unkown, it is most probably because the visitor IP addresses are not being properly propagated to Shlink (due to a reverse proxy missconfiguration or some other network device that's removing that info) or because the GeoLite license key was not provided or is incorrect.

**Reply by GammelSami** [reply]
I don't understand what you mean. The geolocation succeded, the information IS there in the table. The IP addresses are there too.
![grafik](https://user-images.githubusercontent.com/39372285/172021291-55e4a1b7-80c7-421a-9979-e384145395f0.png)
![grafik](https://user-images.githubusercontent.com/39372285/172021720-b74062c2-475b-482e-967b-50727e364883.png)
How can a misconfiguration prevent the output of just one table but all the others are working?

**Reply by acelaya** [reply]
Oh yeah, this looks like a different thing then.

Is the information present in every single row in the database? Do you see "unknown" in every single visit in the frontend? Do you have any kind of proxy cache or response manipulation in between Shlink and the client?

**Reply by GammelSami** [reply]
Every row has an IP address. And every row has a country name. (region and city are empty sometimes)
In frontend every row says "unknown" without exception.
![grafik](https://user-images.githubusercontent.com/39372285/172043458-5eaa7090-0c66-460f-8c04-584027ab1a5e.png)

I'm serving shlink via Plesk's Reverse Proxy feature. In the domain settings nginx-Caching is disabled. And there is a textarea for additional nginx instructions, which is empty.
![grafik](https://user-images.githubusercontent.com/39372285/172043729-63734f27-0b12-44ab-8da5-96a092339616.png)
![grafik](https://user-images.githubusercontent.com/39372285/172045216-4b0bdf37-9224-4508-8f8a-bbc4ccfb50ea.png)

**Reply by acelaya** [reply]
This is super weird. I cannot think of any reason that can be making this happen. Skink servers the visits location info almost verbatim 🤔

I will try to investigate.

Has this happened since the beginning, or did it start to happen with some specific version of the server or client?

**Reply by GammelSami** [reply]
At first I tried a different vServer host provider (also with Plesk). There it worked for the first ≈two days but after that the bug occured. Then I switched the vServer provider (for an other reason) and there the bug occurres since day one.

I just tried the older frontend versions 3.7.0 and 3.5.1 (selfhosted docker) but it still says "unknown".

**Reply by acelaya** [reply]
Can you try bypassing ~traeffik~ plesk's reverse proxy and going directly to the Shlink instance?

**Reply by GammelSami** [reply]
I disabled the firewall, [enabled forwarding](https://docs.docker.com/network/bridge/#enable-forwarding-from-docker-containers-to-the-outside-world) and in the frontend I added the external container port to the server URL. That did not work. ("Oops! Could not connect to this Shlink server.") I don't know what else to try.

But when I run `shlink visit:non-orphan` in terminal I can see the location data like it's in the mysql table.

**Comment by redispade** [comment, marked as answer]
I tried connecting via HTTP directly on docker port bypassing the reverse proxy, and the issue persists

Tried directly on the api:

{
    "visits": {
        "data": [{
            "referer": "https://redacted/",
            "date": "2022-07-01T13:04:17+02:00",
            "userAgent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/103.0.0.0 Safari/537.36",
            "visitLocation": {},
            "potentialBot": false
        }, {
            "referer": "https://www.google.com/",
            "date": "2022-07-01T13:04:14+02:00",
            "userAgent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/102.0.0.0 Safari/537.36",
            "visitLocation": {},
            "potentialBot": false
        }, 

Same thing on HTTP directly on docker container


EDIT: Tried with curl from CLI directly on the docker host and the result is the same. Looks like the locations are on the database, but somehow the REST api doesn't get them. Tried purging the container image, recreating it from scratch(DB is mysql on the host). Still the same. Is there any way I can debug it inside the container?

EDIT2: Switching to the stable tag on docker, fixes the issue. The problem is on the latest.
I am staying on the stable now, but if I can help you find and fix the issue, let me know.

**Reply by acelaya** [reply]
Thanks!

I will check on latest "main". So far I always tested on stable releases and I was never able to reproduce.

**Comment by acelaya** [comment]
I have found the root cause for this. As @redispade found out, no stable release was affected, but it is now fixed on unstable develop branch as well.

**Comment by GammelSami** [comment]
Just updated my Docker containert to 3.4.0, ran `shlink location:visits` and now everything works as it should.
